### PR TITLE
Decorators

### PR DESCRIPTION
**Pull request by Orshahar91**

### CODE REVIEW

File: ex0.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+class InvalidArgumentType(TypeError):
+    """Raise for invalid argument that passes to a function."""
+    __module__ = 'TypeError'
+
+
+def type_check(correct_type):
+    """
+    Decorator factory with a decorator that check if the function receives the corrct type of argument.
+    :param correct_type: The correct type of argument that need to be passed to the function.
+    :return: A decorator.
```
Comment (mattisevitt) [inline ex0.py:13]: This documentation is enough, no need to document all the innner functions.

File: ex0.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
+import functools
+
+
+class InvalidArgumentType(TypeError):
+    """Raise for invalid argument that passes to a function."""
+    __module__ = 'TypeError'
+
+
+def type_check(correct_type):
```
Comment (mattisevitt) [inline ex0.py]: Use type hints and return types -docum
https://www.geeksforgeeks.org/duck-typing-in-python/
https://stackoverflow.com/questions/125367/dynamic-type-languages-versus-static-type-languages
https://docs.python.org/3/library/typing.html
https://realpython.com/lessons/pros-and-cons-type-hints/

File: ex0.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+                                          " and not " + args[0].__class__.__name__)
+            result = function(*args)
+            return result
+        return wrapper
+    return decorator
+
+
+@type_check(int)
+def times(num):
+    return num*2
```
Comment (mattisevitt) [inline ex0.py]: This should not be here.
Don’t use git add -a ever, use git add -p
https://git-scm.com/docs/git-add

File: ex1.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
+def surprise(function):
```
Comment (mattisevitt) [inline ex1.py]: Use type hints and return types -docum
https://www.geeksforgeeks.org/duck-typing-in-python/
https://stackoverflow.com/questions/125367/dynamic-type-languages-versus-static-type-languages
https://docs.python.org/3/library/typing.html
https://realpython.com/lessons/pros-and-cons-type-hints/

File: ex1.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    """
+    Decorator that print surprise instead of the original functionality.
```
Comment (mattisevitt) [inline ex1.py:3]: *that prints

File: ex1.py [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+        return None
+    return wrapper
+
+
+@surprise
+def times(num):
+    return num*2
+
+
+times(2)
```
Comment (mattisevitt) [inline ex1.py:30]: File has no top level environment of the program
https://docs.python.org/3/library/__main__.html

**Review by mattisevitt — CHANGES_REQUESTED** [review]

**Review by mattisevitt — APPROVED** [review]

**Review by talbarfm — APPROVED** [review]

**Pull request merged**